onnxruntime/contrib_ops/cpu/quantization/quant_gemm.cc

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/providers/cpu/math/gemm_base.h"
+#include "core/providers/cpu/math/gemm_helper.h"
+#include "core/providers/cpu/math/matmul_integer_base.h"
+#include "core/quantization/quant_util.h"
+#include "core/util/math_cpuonly.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+class QGemm : protected GemmBase, public MatMulIntegerBase {
+ public:
+  QGemm(const OpKernelInfo& info) : GemmBase(info), MatMulIntegerBase(info) {
+  }
+
+  Status Compute(OpKernelContext* context) const override {
+    const auto* a = context->Input<Tensor>(IN_A);
+    const auto* b = packed_b_ ? nullptr : context->Input<Tensor>(IN_B);
+    const auto& b_shape = b ? b->Shape() : b_shape_;
+
+    const auto* c = context->Input<Tensor>(IN_C);
+    GemmHelper helper(a->Shape(), trans_A_ != CblasNoTrans,
+                      b_shape, trans_B_ != CblasNoTrans,
+                      c != nullptr ? c->Shape() : TensorShape({}));
+    if (!helper.State().IsOK())
+      return helper.State();
+
+    size_t M = SafeInt<size_t>(helper.M());
+    size_t N = SafeInt<size_t>(helper.N());
+    size_t K = SafeInt<size_t>(helper.K());
+
+    //validate scales and zero points
+    const auto* a_zp = context->Input<Tensor>(IN_A_ZERO_POINT);
+    const auto* b_zp = context->Input<Tensor>(IN_B_ZERO_POINT);
+    const auto* y_zp = context->Input<Tensor>(IN_Y_ZERO_POINT);
+    const auto* a_scale = context->Input<Tensor>(IN_A_SCALE);
+    const auto* b_scale = context->Input<Tensor>(IN_B_SCALE);
+    const auto* y_scale = context->Input<Tensor>(IN_Y_SCALE);
+    CheckInputs(a_zp, b_zp, y_zp, a_scale, b_scale, y_scale, helper);
+
+    AllocatorPtr allocator;
+    ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
+
+    BufferUniquePtr a_trans_buffer;
+    const uint8_t* a_data = a->template Data<uint8_t>();
+    if (trans_A_ == CblasTrans) {
+      a_data = quantization::TransPoseInputData(a_data, a_trans_buffer, allocator, K, M);
+    }
+
+    bool b_is_signed;
+    const uint8_t* b_data = nullptr;
+    BufferUniquePtr b_trans_buffer;
+    if (nullptr == b) {
+      b_data = static_cast<const uint8_t*>(packed_b_.get());
+      b_is_signed = b_is_signed_;
+    } else {
+      b_data = static_cast<const uint8_t*>(b->DataRaw());
+      b_is_signed = b->IsDataType<int8_t>();
+      if (trans_B_ == CblasTrans) {
+        b_data = quantization::TransPoseInputData(b_data, b_trans_buffer, allocator, N, K);
+      }
+    }
+
+    auto y = context->Output(OUT_Y, {SafeInt<int64_t>(M), SafeInt<int64_t>(N)});
+    if (M == 0 || N == 0) return Status::OK();
```

Comment:
this should be earlier right?

Reply:
No, need to set the output shape for y even though M == 0 or N == 0